src/instructlab/clickext.py

```diff
@@ -11,11 +12,88 @@
 
 # Third Party
 from click.core import ParameterSource
+from click_didyoumean import DYMGroup
 import click
 
 logger = logging.getLogger(__name__)
 
 
+class LazyEntryPointGroup(DYMGroup):
+    """Lazy load commands from an entry point group
+
+    Entry points are defined in `pyproject.toml`. Example:
+
+        [project.entry-points."instructlab.command.config"]
+        "init" = "instructlab.config.init:init"
+
+    This defines the command `ilab config init` with the function
+    `from instructlab.config.init import init` as click command.
+    """
+
+    def __init__(self, *args, ep_group: str, **kwargs) -> None:
+        super().__init__(*args, **kwargs)
+        self.eps: metadata.EntryPoints = metadata.entry_points(group=ep_group)
+        if not self.eps.names:
+            raise ValueError(f"{ep_group} is empty")
+
+    def list_commands(self, ctx: click.Context) -> list[str]:
+        result = list(super().list_commands(ctx))
+        result.extend(sorted(self.eps.names))
+        return result
+
+    def get_command(self, ctx: click.Context, cmd_name: str) -> click.Command | None:
+        if cmd_name in self.eps.names:
+            cmd = self.eps[cmd_name].load()
+        else:
+            cmd = super().get_command(ctx, cmd_name)
+        if typing.TYPE_CHECKING:
+            assert isinstance(cmd, click.Command) or cmd is None
+        return cmd
+
+
+class ExpandAliasesGroup(LazyEntryPointGroup):
+    """Lazy load commands and aliases from entry point groups"""
+
+    def __init__(self, *args, alias_ep_group: str, **kwargs) -> None:
+        super().__init__(*args, **kwargs)
+        self.alias_eps: metadata.EntryPoints = metadata.entry_points(
+            group=alias_ep_group
+        )
+
+    def get_alias_info(self, cmd_name: str) -> tuple[str, str]:
+        ep: metadata.EntryPoint = self.alias_eps[cmd_name]
+        # assume that the second item of the module name is the group
+        return ep.module.split(".", 3)[1], ep.attr
+
+    def get_command(self, ctx: click.Context, cmd_name: str) -> click.Command | None:
+        if cmd_name in self.alias_eps.names:
+            cmd = self.alias_eps[cmd_name].load()
+            if typing.TYPE_CHECKING:
```

Comment:
Why would we need to only run this assertion during the type-checking process, would we not want it to be a part of the regular execution? Unless I'm mistaken, it seems like it would satisfy the type-checker, but then enable the problem to exist at runtime.

Reply:
This is the standard way to assist type checker with additional type hints. The `load` function returns `typing.Any`, but we know it's a command instance.

Reply:
I see, that makes sense now.

Reply:
After 0.18.0, I'm going to add some docs about typing and type annotations. Eventually I like to have the entire code typed. It helps with readability, code completion in IDEs, and most importantly to find bugs.

Reply:
That will be really helpful to have. I'm always a +1 on adding strong typing.